src/compiler/checker.ts

```diff
@@ -16871,23 +16848,60 @@ namespace ts {
             return isLengthPushOrUnshift || isElementAssignment;
         }
 
-        function maybeTypePredicateCall(node: CallExpression) {
-            const links = getNodeLinks(node);
-            if (links.maybeTypePredicate === undefined) {
-                links.maybeTypePredicate = getMaybeTypePredicate(node);
+        function isDeclarationWithExplicitTypeAnnotation(declaration: Declaration | undefined) {
+            return !!(declaration && (
+                declaration.kind === SyntaxKind.VariableDeclaration || declaration.kind === SyntaxKind.Parameter ||
+                declaration.kind === SyntaxKind.PropertyDeclaration || declaration.kind === SyntaxKind.PropertySignature) &&
+                (declaration as VariableDeclaration | ParameterDeclaration | PropertyDeclaration | PropertySignature).type);
+        }
+
+        function getExplicitTypeOfSymbol(symbol: Symbol) {
+            return symbol.flags & (SymbolFlags.Function | SymbolFlags.Method | SymbolFlags.Class | SymbolFlags.ValueModule) ||
+                symbol.flags & (SymbolFlags.Variable | SymbolFlags.Property) && isDeclarationWithExplicitTypeAnnotation(symbol.valueDeclaration) ?
+                getTypeOfSymbol(symbol) : undefined;
+        }
+
+        function getTypeOfDottedName(node: Expression) {

```

Comment:
This.... deserves remarks on why we can't just call `checkPropertyAccessExpression` or even `checkExpression` (at a glace, `getTypeOfDottedName` in no way indicates that it's only looking up declared types and refraining from resolving anything else).

Reply:
Agreed.

Reply:
Actually, there's already a comment to that effect, but I'll add some more wording.

Reply:
Preferably on the function jsdoc style, rather than in the body. At least named as-is, it's tempting to use elsewhere.